Public/Src/Utilities/Storage/ChangeTracking/FileChangeProcessor.cs

```diff
@@ -52,29 +66,42 @@ public void Subscribe(IFileChangeTrackingObserver observer)
             JournalProcessingStatistics.LogMessage logMessage = null,
             JournalProcessingStatistics.LogStats logStats = null)
         {
-            if (m_scanningJournalResult != null)
-            {
-                return m_scanningJournalResult;
-            }
+            InitObservers();
 
-            // Initialize observers.
-            foreach (var unsubscriber in m_unsubscribers)
+            // Input change list needs to be processed first because file change tracker
+            // makes a completion for the observers currently.
+            // TODO: Make a better design for many-to-many observable-observer relations.
+            m_inputChangeList?.ProcessChanges();
+            var scanningJournalResult = m_fileChangeTracker.TryProcessChanges(timeLimit);
+
+            UnsubscribeObservers();
+
+            var journalStatistics = new JournalProcessingStatistics(scanningJournalResult);
+            journalStatistics.Log(m_loggingContext, logMessage, logStats);
+
+            return scanningJournalResult;
+        }
+
+        private void InitObservers()
+        {
+            // Only file change tracker's observers need to be initialized.
+            foreach (var unsubscriber in m_fileChangeTrackerUnsubscribers)

```

Comment:
>m_fileChangeTrackerUnsubscribers [](start = 41, length = 32)

Now the name is  a little funny. Unsubscriber sounds like it is for cleanup, but used for initialization here... should it just be `m_fileChangeTrackers`?

Reply:
The unsubscriber contains the observers. Yes, the name is a bit funny. I'm just following the design pattern in MSDN.

---
In reply to: [313497206](https://github.com/microsoft/BuildXL/pull/726#discussion_r313497206) [](ancestors = 313497206)